### ISSUE TITLE
CDK8s 2.x.x,  RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by flashflexpro** [cdk8s-team/cdk8s#951]
all cdk8s:"1.x.x" working great,  this only happens with   `  "dependencies": {   "cdk8s": "^2.x.x",   ... `

/home/ubuntu/IdeaProjects/cdk8s/node_modules/cdk8s/lib/_resolve.js:5
function resolve(value) {
                ^

RangeError: Maximum call stack size exceeded
    at resolve (/home/ubuntu/IdeaProjects/cdk8s/node_modules/cdk8s/lib/_resolve.js:5:17)
    at resolve (/home/ubuntu/IdeaProjects/cdk8s/node_modules/cdk8s/lib/_resolve.js:27:21)
    at resolve (/home/ubuntu/IdeaProjects/cdk8s/node_modules/cdk8s/lib/_resolve.js:27:21)
    at resolve (/home/ubuntu/IdeaProjects/cdk8s/node_modules/cdk8s/lib/_resolve.js:27:21)
    at resolve (/home/ubuntu/IdeaProjects/cdk8s/node_modules/cdk8s/lib/_resolve.js:27:21)
    at resolve (/home/ubuntu/IdeaProjects/cdk8s/node_modules/cdk8s/lib/_resolve.js:27:21)
    at resolve (/home/ubuntu/IdeaProjects/cdk8s/node_modules/cdk8s/lib/_resolve.js:27:21)
    at resolve (/home/ubuntu/IdeaProjects/cdk8s/node_modules/cdk8s/lib/_resolve.js:27:21)
    at resolve (/home/ubuntu/IdeaProjects/cdk8s/node_modules/cdk8s/lib/_resolve.js:27:21)
    at resolve (/home/ubuntu/IdeaProjects/cdk8s/node_modules/cdk8s/lib/_resolve.js:27:21)
Error: command "node main.js " at /home/ubuntu/IdeaProjects/cdk8s returned a non-zero exit code 1
    at ChildProcess.<anonymous> (/home/ubuntu/IdeaProjects/cdk8s/node_modules/cdk8s-cli/lib/util.js:50:27)
command "node main.js " at /home/ubuntu/IdeaProjects/cdk8s returned a non-zero exit code 1

    at Object.onceWrapper (node:events:510:26)
    at ChildProcess.emit (node:events:390:28)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:290:12)

> Task :cdk8s_build FAILED

Execution failed for task ':cdk8s_build'.
> Process 'command '/home/ubuntu/IdeaProjects/cdk8s/.gradle/npm/npm-v8.11.0/bin/npm'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
`

**Comment by iliapolo** [comment]
@flashflexpro Can you please attach the code you're using as well?

**Comment by diefans** [comment]
I am facing the same issue while trying to define a `KubeClusterRoleBinding`:

```
    const serviceAccount = new KubeServiceAccount(this, "fluentd-svc-account", {
      metadata: {
        name: "fluentd",
        namespace: "kube-system",
      }
    });

    const clusterRole = new KubeClusterRole(this, "fluentd-cluster-role", {
      metadata: {
        name: "fluentd",
      },
      rules: [
        {
          apiGroups: [""],
          resources: [ "pods", "namespaces" ],
          verbs: [ "get", "list", "watch" ]
        }
      ]
    });

    new KubeClusterRoleBinding(this, "fluentd-cluster-role-binding", {
      metadata: {
        name: "fluentd",
      },
      roleRef: clusterRole,
      subjects: [serviceAccount],
    });
```

**Comment by diefans** [comment]
cdk8s version is v2.3.74

**Comment by diefans** [comment]
I can avoid this issue by not referencing any object in the role binding:
 ``` typescript
    new KubeClusterRoleBinding(this, "fluentd-cluster-role-binding", {
      metadata: {
        name: "fluentd",
      },
      roleRef: {
        kind: "ClusterRole",
        name: "fluentd",
        apiGroup: "rbac.authorization.k8s.io"
      },
      subjects: [{
        kind: "ServiceAccount",
        name: "fluentd",
        namespace: "kube-system"
      }],
    });
```
Question is: is this the intended way?

**Comment by jakubknejzlik** [comment]
Hello, I just hit same error with this sample code:
```typescript
const configMap = new kplus.ConfigMap(chart, 'ConfigMap', {
  data: {FOO:'blah'},
});
const deployment = new kplus.Deployment(this, 'BookstoreWebsite', {
  containers: [{ 
    image: 'node', 
    // without this line it works fine
    envFrom: [new kplus.EnvFrom(configMap)] 
  }],
});
new kplus.HorizontalPodAutoscaler(this, 'DeploymentAutoscaler', {
  target: deployment,
  maxReplicas: 20,
  minReplicas: 1,
  metrics: [kplus.Metric.resourceCpu(kplus.MetricTarget.averageUtilization(70))],
});
```
Versions:
```
├── @types/jest@26.0.24
├── @types/node@14.18.26
├── cdk8s-cli@2.0.95
├── cdk8s-plus-25@2.0.19
├── cdk8s@2.5.37
├── constructs@10.1.152
├── jest@26.6.3
├── ts-jest@26.5.6
└── typescript@4.8.2
```

**Comment by ibarrick** [comment]
I've run into the same issue with HorizontalPodAutoscaler. HPA doesn't appear to support another way of specifying the target.

**Comment by garyy7811** [comment]
the problem I encountered is caused by imported classes can have a self reference when doing `Object.entries()` o when [trying to resolve ](https://github.com/cdk8s-team/cdk8s-core/blob/2a058727abf2aa76b0af0b235174b23e3483a28c/src/_resolve.ts#L31)

So import cli need to add a test to make sure generated class will not reference to self!!

**Comment by iliapolo** [comment]
Hey Folks. There seem to several things going on in this issue. I'll try to clear everything up. 

----

@diefans [wrote](https://github.com/cdk8s-team/cdk8s/issues/951#issuecomment-1204038507):

> Question is: is this the intended way?

Yes. In general, when using the L1 layer (i.e `Kube* constructs`), you should never pass objects around between resources. All L1 constructs expect primitives or structs, the fact this passes compilation is an unfortunate side-effect of TypeScript structural typing. That is, the `roleRef` property expects three properties: `kind`, `name` and `apiGroup`; In cdk8s every single resource has these properties, so incidentally every single resource "implements" a `roleRef` interface. But this is not a correct API usage. In languages without structural typing like Java for example, this would have failed compilation.

I agree this is not an ideal experience, i'll take a note to see if we can add some runtime checks to make this misuse more clear.

----

@njlaw [wrote](https://github.com/cdk8s-team/cdk8s/issues/951#issuecomment-1255264020):

> But this does not even though I am referencing the name attributes on the KubeRole and KubeServiceAccount.

Do you mean you still get the error even though you are only passing the name and not the object? This shouldn't happen, and I couldn't reproduce it. Please let us know if this is what you are seeing.

----

@jakubknejzlik @ibarrick the error you are seeing with HPA was fixed in this [PR](https://github.com/cdk8s-team/cdk8s-plus/pull/1438). I verified that running the code you provided works with the latest version. Please upgrade and let us know if you still see a problem,

----

@garyy7811 [wrote](https://github.com/cdk8s-team/cdk8s/issues/951#issuecomment-1407462496):

> the problem I encountered is caused by imported classes can have a self reference when doing Object.entries() o when [trying to resolve](https://github.com/cdk8s-team/cdk8s-core/blob/2a058727abf2aa76b0af0b235174b23e3483a28c/src/_resolve.ts#L31)

Lets track this problem [here](https://github.com/cdk8s-team/cdk8s-cli/issues/793), as it sounds like a different use-case (albeit same error). Can you also provide the specific CRD you trying to import? 

----

So, to summarize...the problems mentioned here were either fixed or are a result of incorrect API usage. I'll be closing this issue out soon. As for the CRD import issue, we will track it the [this](https://github.com/cdk8s-team/cdk8s-cli/issues/793) issue.

**Comment by njlaw** [comment]
> > But this does _not_ even though I am referencing the name attributes on the KubeRole and KubeServiceAccount.
> 
> Do you mean you still get the error even though you are only passing the name and not the object? This shouldn't happen, and I couldn't reproduce it. Please let us know if this is what you are seeing.

Thank you for the explanation.  What I am seeing is the expected behavior: it does work if I just pass the name.

**Comment by iliapolo** [comment]
Closing this out. The [CRD import](https://github.com/cdk8s-team/cdk8s-cli/issues/793) issue has also been resolved.
